python/tests/stress/arcticdb/version_store/test_mem_leaks.py

```diff
@@ -209,3 +383,201 @@ def proc_to_examine():
+            what we must exclude from calculation
+        """
+        (lib, df) = library_with_symbol
+        mem_query(lib, df)
+
+
+    @MEMRAY_TESTS_MARK
+    @pytest.mark.limit_leaks(location_limit="25 KB", filter_fn=is_relevant)
+    def test_mem_leak_querybuilder_read_batch_memray(library_with_symbol):
+        """
+            Test to capture memory leaks >= of specified number
+
+            NOTE: we could filter out not meaningful for us stackframes
+            in future if something outside of us start to leak using
+            the argument "filter_fn" - just add to the filter function
+            what we must exclude from calculation
+        """
+        (lib, df) = library_with_symbol
+        mem_query(lib, df, read_batch=True)
+
+
+    @MEMRAY_TESTS_MARK
+    @pytest.mark.limit_memory("490 MB")
+    def test_mem_limit_querybuilder_read_memray(library_with_symbol):
+        """
+            The fact that we do not leak memory does not mean that we
+            are efficient on memory usage. This test captures the memory usage
+            and limits it, so that we do not go over it (unless fro good reason)
+            Thus if the test fails then perhaps we are using now more memory than
+            in the past
+        """
+        (lib, df) = library_with_symbol
+        mem_query(lib, df)
+
+    @MEMRAY_TESTS_MARK
+    @pytest.mark.limit_memory("490 MB")
+    def test_mem_limit_querybuilder_read_batch_memray(library_with_symbol):
+        """
+            The fact that we do not leak memory does not mean that we
+            are efficient on memory usage. This test captures the memory usage
+            and limits it, so that we do not go over it (unless fro good reason)
+            Thus if the test fails then perhaps we are using now more memory than
+            in the past
+        """
+        (lib, df) = library_with_symbol
+        mem_query(lib, df, True)
+
+
+    @pytest.fixture
+    def library_with_big_symbol(arctic_library_lmdb) -> Generator[Library, None, None]:
+        """
+            As memray instruments memory, we need to take out 
+            everything not relevant from mem leak measurement out of 
+            test, so it works as less as possible
+        """
+        lib: Library = arctic_library_lmdb
+        df = generate_big_dataframe(300000)
+        lib.write(symbol, df)
+        del df
+        yield lib
+
+
+    @MEMRAY_TESTS_MARK
+    @pytest.mark.limit_leaks(location_limit="30 KB", filter_fn=is_relevant)
+    def test_mem_leak_read_all_arctic_lib_memray(library_with_big_symbol):
+        """
+            This is a new version of the initial test that reads the whole
+            big dataframe in memory
+        """
+
+        print("Test starting")
+        st = time.time()
+        lib: Library = library_with_big_symbol
+        data = lib.read(symbol).data
+        del data
+        print("Test took :", time.time() - st)
+
+        print("Sleeping for 10 secs")
```

Comment:
Why?

Reply:
I do not think gc.collect() is synchronous task. Or at least my observations with psutil library are different. Memory is not released synchronous to the call or at least not all of it. Thus the decision for flocking sleep. 
That gives oppertunity additional release process to kick in. 
Again this is from observation with other approach (with psutils, prior to memray) where I could not make it relativly stable unless sleep is introduced. Again this sleep is only for those kind of tests it is not clobal one, thus it can stay. If removed I do not know what the results could be

Reply:
There's nothing in the [docs](https://docs.python.org/3/library/gc.html#gc.collect) suggesting this is asynchronous. What `psutil` stats were you comparing to?

Reply:
Well reallity is different.

See how psutils sees memory without sleep and with sleep. Ovviously there are processes that require 'calm' process in order for external tools on which psutility is based to measure properly memory.

Thus having NO sleep it is rather road to flaky tests - I have observed that already with standard approach. Is it going to repeat itself with memray tests I do not know, but it is better to pay the price of waiting. Remember patience is a virtue ;-) In our case it might produce lots of flakyness and for what?


-----------------------------------------------------------------------------
  WITHOUT WAITS 10 SEC
-----------------------------------------------------------------------------

Process initial RSS  1700.16MB/[1782743040] 
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[0]    : Process added (or if number is negative - process released)  -36.97MB/[-38768640] .  Avail memory:  43992.38MB/[46129352704]  Used Mem: 8.7%
  Overall stats : Process growth since start  -36.97MB/[-38768640]  AVG growth per iter  -36.97MB/[-38768640.0] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  -36.97MB/[-38768640] 
  Number of times there was 50% drop in memory: 0
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[1]    : Process added (or if number is negative - process released)  0.17MB/[180224] .  Avail memory:  44074.61MB/[46215573504]  Used Mem: 8.5%
  Overall stats : Process growth since start  -36.80MB/[-38588416]  AVG growth per iter  -18.40MB/[-19294208.0] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  -36.80MB/[-38588416] 
  Number of times there was 50% drop in memory: 1
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[2]    : Process added (or if number is negative - process released)  13.77MB/[14442496] .  Avail memory:  44088.59MB/[46230241280]  Used Mem: 8.5%
  Overall stats : Process growth since start  -23.03MB/[-24145920]  AVG growth per iter  -7.68MB/[-8048640.0] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  -23.03MB/[-24145920] 
  Number of times there was 50% drop in memory: 2
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[3]    : Process added (or if number is negative - process released)  3.79MB/[3977216] .  Avail memory:  44066.72MB/[46207307776]  Used Mem: 8.5%
  Overall stats : Process growth since start  -19.23MB/[-20168704]  AVG growth per iter  -4.81MB/[-5042176.0] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  -19.23MB/[-20168704] 
  Number of times there was 50% drop in memory: 3
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[4]    : Process added (or if number is negative - process released)  43.75MB/[45879296] .  Avail memory:  44020.70MB/[46159044608]  Used Mem: 8.6%
  Overall stats : Process growth since start  24.52MB/[25710592]  AVG growth per iter  4.90MB/[5142118.4] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  24.52MB/[25710592] 
  Number of times there was 50% drop in memory: 3
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[5]    : Process added (or if number is negative - process released)  103.92MB/[108969984] .  Avail memory:  43923.30MB/[46056919040]  Used Mem: 8.8%
  Overall stats : Process growth since start  128.44MB/[134680576]  AVG growth per iter  21.41MB/[22446762.666666668] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  128.44MB/[134680576] 
  Number of times there was 50% drop in memory: 3
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[6]    : Process added (or if number is negative - process released)  -22.30MB/[-23379968] .  Avail memory:  43941.50MB/[46076002304]  Used Mem: 8.8%
  Overall stats : Process growth since start  106.14MB/[111300608]  AVG growth per iter  15.16MB/[15900086.857142856] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  128.44MB/[134680576] 
  Number of times there was 50% drop in memory: 3
Starting watched code ...........
Lets pause for 0 secs so GC to kick in
Iter No[7]    : Process added (or if number is negative - process released)  -138.72MB/[-145457152] .  Avail memory:  44065.91MB/[46206459904]  Used Mem: 8.5%
  Overall stats : Process growth since start  -32.57MB/[-34156544]  AVG growth per iter  -4.07MB/[-4269568.0] 
  Minimum growth so far:  -36.97MB/[-38768640] 
  Maximum growth so far:  128.44MB/[134680576] 
  Number of times there was 50% drop in memory: 4
  

-----------------------------------------------------------------------------
  WITH WAITS 10 SEC
-----------------------------------------------------------------------------
Process initial RSS  1734.86MB/[1819131904] 
Starting watched code ...........
Lets pause for 10 secs so GC to kick in
Iter No[0]    : Process added (or if number is negative - process released)  45.19MB/[47386624] .  Avail memory:  43974.38MB/[46110478336]  Used Mem: 8.7%
  Overall stats : Process growth since start  45.19MB/[47386624]  AVG growth per iter  45.19MB/[47386624.0] 
  Minimum growth so far:  45.19MB/[47386624] 
  Maximum growth so far:  45.19MB/[47386624] 
  Number of times there was 50% drop in memory: 0
Starting watched code ...........
Lets pause for 10 secs so GC to kick in
Iter No[1]    : Process added (or if number is negative - process released)  -7.00MB/[-7344128] .  Avail memory:  43968.44MB/[46104248320]  Used Mem: 8.7%
  Overall stats : Process growth since start  38.19MB/[40042496]  AVG growth per iter  19.09MB/[20021248.0] 
  Minimum growth so far:  38.19MB/[40042496] 
  Maximum growth so far:  45.19MB/[47386624] 
  Number of times there was 50% drop in memory: 0
Starting watched code ...........
Lets pause for 10 secs so GC to kick in
Iter No[2]    : Process added (or if number is negative - process released)  -82.82MB/[-86847488] .  Avail memory:  44054.60MB/[46194597888]  Used Mem: 8.6%
  Overall stats : Process growth since start  -44.64MB/[-46804992]  AVG growth per iter  -14.88MB/[-15601664.0] 
  Minimum growth so far:  -44.64MB/[-46804992] 
  Maximum growth so far:  45.19MB/[47386624] 
  Number of times there was 50% drop in memory: 1
Starting watched code ...........
Lets pause for 10 secs so GC to kick in
Iter No[3]    : Process added (or if number is negative - process released)  88.26MB/[92545024] .  Avail memory:  43922.98MB/[46056579072]  Used Mem: 8.8%
  Overall stats : Process growth since start  43.62MB/[45740032]  AVG growth per iter  10.91MB/[11435008.0] 
  Minimum growth so far:  -44.64MB/[-46804992] 
  Maximum growth so far:  45.19MB/[47386624] 
  Number of times there was 50% drop in memory: 1
Starting watched code ...........
Lets pause for 10 secs so GC to kick in
Iter No[4]    : Process added (or if number is negative - process released)  -82.00MB/[-85979136] .  Avail memory:  44019.06MB/[46157332480]  Used Mem: 8.6%
  Overall stats : Process growth since start  -38.38MB/[-40239104]  AVG growth per iter  -7.67MB/[-8047820.8] 
  Minimum growth so far:  -44.64MB/[-46804992] 
  Maximum growth so far:  45.19MB/[47386624] 
  Number of times there was 50% drop in memory: 2
Starting watched code ...........
Lets pause for 10 secs so GC to kick in
Iter No[5]    : Process added (or if number is negative - process released)  42.20MB/[44249088] .  Avail memory:  43991.38MB/[46128308224]  Used Mem: 8.7%
  Overall stats : Process growth since start  3.82MB/[4009984]  AVG growth per iter  0.64MB/[668330.6666666666] 
  Minimum growth so far:  -44.64MB/[-46804992] 
  Maximum growth so far:  45.19MB/[47386624] 
  Number of times there was 50% drop in memory: 2
Starting watched code ...........
Lets pause for 10 secs so GC to kick in
Iter No[6]    : Process added (or if number is negative - process released)  -58.56MB/[-61403136] .  Avail memory:  44048.85MB/[46188564480]  Used Mem: 8.6%
  Overall stats : Process growth since start  -54.73MB/[-57393152]  AVG growth per iter  -7.82MB/[-8199021.714285715] 
  Minimum growth so far:  -54.73MB/[-57393152] 
  Maximum growth so far:  45.19MB/[47386624] 
  Number of times there was 50% drop in memory: 3
Starting watched code ...........
Lets pause for 10 secs so GC to kick in